src/main/java/io/cdap/plugin/spark/HierarchyToRelationalUtils.java

```diff
@@ -0,0 +1,211 @@
+    leafRecords.registerTempTable(SQLUtils.NAME_LEAVES_TEMP_TABLE);
+    // TODO - required in order to be able to iterate - can pass data and sqlContext otherwise
+    List<Row> rows = uniqueRecords.collectAsList();
+
+    // Empty Dataset to hold all the results
+    Dataset<Row> results = sqlContext.createDataFrame(
+      sparkExecutionPluginContext.getSparkContext().emptyRDD(), DataFrames.toDataType(outputSchema));
+    // Root does not exists as separate value - this will generate it
+    results = results.union(generateTreeRoot(data, config));
+    // Generate tree for each unique row
+    for (Row row : rows) {
+      results = results.union(generateQuery(row, data, sqlContext));
+    }
+    // sets leaves indicator
+    results.registerTempTable(SQLUtils.NAME_RESULTS_TEMP_TABLE);
+    return sqlContext.sql(SQLUtils.sqlStatementForSettingLeafsValue(config));
+  }
+
+  /**
+   * Generate dataset containing all subtree details for a given row
+   *
+   * @param parentRow  row for which to extract the subtree
+   * @param data       reference to input dataset
+   * @param sqlContext reference to {@link SQLContext}
+   * @return dataset reference with all subtree rows for a given row
+   * @throws Exception raises exception if max depth is exceeded
+   */
+  public Dataset<Row> generateQuery(Row parentRow, Dataset<Row> data, SQLContext sqlContext)
+    throws Exception {
+    long dataFrameCount = 1;
+    int count = 1;
+
+    // make parent id sql ready based on type
+    String parentId;
+    String parentIdValue = String.valueOf(parentRow.get(0));
+    if (NumberUtils.isNumber(parentIdValue)) {
+      parentId = parentIdValue;
+    } else {
+      parentId = String.format("'%s'", parentRow.getString(0));
+    }
+
+    // TODO - will require all input fields and config of parent fields in order to swap parent id and parent field
+    // ex: changing from 11,13,Milk,Chocolate Milk with 3,13,Dairy,Chocolate Milk
+    // currently we have only parent id
+    mapSelfToTreeRoot(parentRow, data, config);
+
+    //TODO - move to centralized location (waiting on reply)
+    String parentProductField = "ParentProduct";
+    String childProductField = "ChildProduct";
+    String otherFields = "Supplier, Sales";
+
+    Dataset<Row> dataFrameSeed = data.sqlContext()
+      .sql(SQLUtils.startQueryForRecursive(config, SQLUtils.NAME_HIERARCHY_TABLE, parentId));
+    dataFrameSeed.registerTempTable(String.format("vt_%s_seed_0", parentId));
+
+    Dataset<Row> branchRoot = data.sqlContext().sql(String.format("select %s from vt_%s_seed_0", parentProductField,
+                                                                  parentId));
+    if (branchRoot.count() > 0) {
+      // TODO - move to centralized location (waiting on reply)
+      parentProductField = String.format("'%s' as %s", branchRoot.first().get(0), parentProductField);
+    }
+
+    while (dataFrameCount != 0) {
+      if (count == Integer.parseInt(config.getMaxDepthField())) {
```

Comment:
if max depth is 1, we'll throw an exception because count is initialized to 1. Is that expected?
Have you tested with non default max depth configured?